subsys/net/lib/rest_client/src/rest_client.c

```diff
@@ -0,0 +1,360 @@
+		LOG_DBG("Partial data received(%zd bytes)", rsp->data_len);
+	} else if (final_data == HTTP_DATA_FINAL) {
+		if (!rest_ctx) {
+			LOG_WRN("REST context not provided");
+			return;
+		}
+		rest_ctx->http_status_code = rsp->http_status_code;
+		rest_ctx->response_len = rsp->content_length;
+
+		LOG_DBG("HTTP: All data received (content/total: %d/%d), status: %u %s",
+			rsp->content_length,
+			rest_ctx->total_response_len,
+			rsp->http_status_code,
+			log_strdup(rsp->http_status));
+	}
+}
+
+static int rest_client_sckt_tls_setup(int fd, const char *const tls_hostname,
+				      const sec_tag_t sec_tag, int tls_peer_verify)
+{
+	int err;
+	int verify = TLS_PEER_VERIFY_REQUIRED;
+	const sec_tag_t tls_sec_tag[] = {
+		sec_tag,
+	};
+	uint8_t cache;
+
+	if (tls_peer_verify == TLS_PEER_VERIFY_REQUIRED ||
+	    tls_peer_verify == TLS_PEER_VERIFY_OPTIONAL ||
+	    tls_peer_verify == TLS_PEER_VERIFY_NONE) {
+		verify = tls_peer_verify;
+	}
+
+	err = setsockopt(fd, SOL_TLS, TLS_PEER_VERIFY, &verify, sizeof(verify));
+	if (err) {
+		LOG_ERR("Failed to setup peer verification, error: %d", errno);
+		return err;
+	}
+
+	err = setsockopt(fd, SOL_TLS, TLS_SEC_TAG_LIST, tls_sec_tag, sizeof(tls_sec_tag));
+	if (err) {
+		LOG_ERR("Failed to setup TLS sec tag, error: %d", errno);
+		return err;
+	}
+
+	if (IS_ENABLED(CONFIG_REST_CLIENT_SCKT_TLS_SESSION_CACHE_IN_USE)) {
+		cache = TLS_SESSION_CACHE_ENABLED;
+	} else {
+		cache = TLS_SESSION_CACHE_DISABLED;
+	}
+
+	err = setsockopt(fd, SOL_TLS, TLS_SESSION_CACHE, &cache, sizeof(cache));
+	if (err) {
+		LOG_ERR("Unable to set session cache, errno %d", errno);
+		return err;
+	}
+
+	if (tls_hostname) {
+		err = setsockopt(fd, SOL_TLS, TLS_HOSTNAME, tls_hostname, strlen(tls_hostname));
+		if (err) {
+			LOG_ERR("Failed to setup TLS hostname, error: %d", errno);
+			return err;
+		}
+	}
+	return 0;
+}
+
+static int rest_client_sckt_timeouts_set(int fd)
+{
+	int err;
+	struct timeval timeout = { 0 };
+
+	if (CONFIG_REST_CLIENT_SCKT_SEND_TIMEOUT > -1) {
+		/* Send TO also affects TCP connect */
+		timeout.tv_sec = CONFIG_REST_CLIENT_SCKT_SEND_TIMEOUT;
+		err = setsockopt(fd, SOL_SOCKET, SO_SNDTIMEO, &timeout, sizeof(timeout));
+		if (err) {
+			LOG_ERR("Failed to set socket send timeout, error: %d", errno);
+			return err;
+		}
+	}
+
+	if (CONFIG_REST_CLIENT_SCKT_RECV_TIMEOUT > -1) {
+		timeout.tv_sec = CONFIG_REST_CLIENT_SCKT_RECV_TIMEOUT;
+		err = setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
+		if (err) {
+			LOG_ERR("Failed to set socket recv timeout, error: %d", errno);
+			return err;
+		}
+	}
+	return 0;
+}
+
+static int rest_client_sckt_connect(int *const fd, const char *const hostname,
+				     const uint16_t port_num,
+				     const sec_tag_t sec_tag,
+				     int tls_peer_verify)
+{
+	int ret;
+	struct addrinfo *addr_info;
+	char peer_addr[INET_ADDRSTRLEN];
+	const char *const connect_addr = hostname;
+
+	struct addrinfo hints = {
+		.ai_family = AF_INET,
```

Comment:
How about IPv6 support, i.e. using AF_UNSPEC? This would maybe need to have a flag in the client context, though.

Reply:
A good point, IPv6 not currently supported. I'll change this to use AF_UNSPEC but left the force of using certain family to limitations.